ccm_web/server/src/api/api.service.ts

```diff
@@ -120,6 +128,35 @@ export class APIService {
     return await sectionHandler.enrollUsers(sectionUsers)
   }
 
+  async enrollSectionExternalUsers (user: User, sectionId: number, sectionUsers: SectionExternalUserDto[]): Promise<string> {
+    // Create all requested users, noting failures
+    const adminRequestor = this.canvasService.createRequestorForAdmin('/api/v1/')
+    const adminHandler = new AdminApiHandler(adminRequestor, user.loginId)
+    const canvasConfig = this.configService.get('canvas', { infer: true })
+    const createUserResponses = await adminHandler.createExternalUsers(sectionUsers, canvasConfig.newUserAccountID)
+    const newUsers = createUserResponses.filter(response => !isAPIErrorData(response)) as CanvasUser[]
+
+    // Invite only new users
+    const inviteResults: string | null = await this.invitationService.sendInvitations(newUsers)
+
+    // Enroll all users
+    const requestor = await this.canvasService.createRequestorForUser(user, '/api/v1/')
+    const sectionHandler = new SectionApiHandler(requestor, sectionId)
+    const userEnrollments = await sectionHandler.enrollUsers(sectionUsers)
+
+    return JSON.stringify({
```

Comment:
Would it be possible to not `JSON.stringify` here? I don't think we're doing that anywhere else. I'd prefer to create an interface for the result object you're composing.

Once this is merged, I may have some more thoughts on the structure here. I think from a requirements perspective, we should be able to what new users were created, and whether any failures occurred. It'd be nice if the UI didn't have to do too much data wrangling to get those answers.

Reply:
I didn't think it was important enough to define a new enrollment/invitation type to return from `enrollSectionExternalUsers()`, so I decided to return a simple string, containing JSON.  It looks like using `JSON.stringify()` here is the only way to accomplish that.  I tried changing the `return` statement to use `String()` instead, but the result that's returned when using that is `[object Object]`.

Maybe I'll try returning just `Object` and see whether that's automatically converted to JSON.

Reply:
Yes, it was as simple as using `object` as the return type.  I guess if it's necessary/desirable to define a new return type more specific than `object`, I can do that.

Reply:
I believe we are missing out on type-related benefits if you just return `object` or `string`. By specifying the return type, you will know when you change any method internals if the type has changed consequently.

Upon reflection, would it be possible for this method to interpret and pre-package the results so the frontend would know a) whether the full three-part operation was successful for a particular user, and b) if the user already existed in Canvas or not. b) may not be necessary, it depends on the UI requirements, but I think it would be beneficial if we could report a partial error. e.g. a table with the users that failed to be added and the reason why if feasible.

Reply:
I'll look into that.  I also see that eslint really doesn't like that my code returns the `object` type and we want to keep that happy.  So, I guess that alone is a good enough reason to return a better type.